lib/flight_scheduler/path_generator.rb

```diff
@@ -0,0 +1,129 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, EITHER EXPRESS OR
+# IMPLIED INCLUDING, WITHOUT LIMITATION, ANY WARRANTIES OR CONDITIONS
+# OF TITLE, NON-INFRINGEMENT, MERCHANTABILITY OR FITNESS FOR A
+# PARTICULAR PURPOSE. See the Eclipse Public License 2.0 for more
+# details.
+#
+# You should have received a copy of the Eclipse Public License 2.0
+# along with FlightSchedulerController. If not, see:
+#
+#  https://opensource.org/licenses/EPL-2.0
+#
+# For more information on FlightSchedulerController, please visit:
+# https://github.com/openflighthpc/flight-scheduler-controller
+#==============================================================================
+
+require 'etc'
+
+module FlightScheduler
+  # NOTE: PathGenerator has been deliberately implemented without knowledge of
+  # the data model. This is to prevent coupling to the current implementation
+  # of Task-Job relationship
+  class PathGenerator
+    NUMERIC_KEYS = {
+      'a' => 'The current index when running an array task, otherwise 0',
+      # TODO: This can't be implemented affectively as the nodes are allocated
+      # to individual array tasks and not the job. This makes determining the
+      # node indexing either inconsistent or impossible to determine
+      #
+      # The allocation behaviour is likely to change TBD. Consider revisiting
+      # 'n' => 'The relative ID of the node within the job',
+    }
+    ALPHA_KEYS = {
+      'A' => "The ID of the associate job when running an array task, otherwise empty string",
+      'j' => 'The ID of non array task jobs, otherwise empty string',
+      'N' => 'The current node name',
+      'u' => 'The user name',
+      'x' => 'The job name'
+    }
+
+    DESC = [
+      'The paths may contain various special characters which will be replaced:',
+      '',
+      *NUMERIC_KEYS.map { |k, d| " * `%\\d*#{k}`: #{d}" },
+      *ALPHA_KEYS.map { |k, d| " * `%#{k}`: #{d}" },
+      " * `%%`: Escape a literal percent character '%', instead of a special directive",
+      ' * `%<char>`: All other characters form an invalid replacement'
+    ].join("\n")
+
+    ALL_CHARS = [*ALPHA_KEYS.keys, *NUMERIC_KEYS.keys.join('')]
+
+    # NOTE: The \\d is converted to \d via string interpolation before typecasting to regex
+    PCT_REGEX = Regexp.new "%+\\d*[#{ALL_CHARS.join('')}]?"
+    PAD_REGEX = /(\d*).\Z/
+    GENERAL_REGEX = /%+\d*[^%]?/
+
+    def self.valid?(path)
+      path.scan(GENERAL_REGEX).all? do |part|
+        if part.count('%').even?
+          true
+        elsif NUMERIC_KEYS[part[-1]]
+          true
+        elsif ALPHA_KEYS[part[-1]]
+          PAD_REGEX.match(part).captures[0].empty?
+        else
+          false
+        end
+      end
+    end
+
+    attr_reader :node, :job, :task
+
+    def initialize(node:, job:, task: nil)
+      @node = node
+      @job = job
+      @task = task
+    end
+
+    def pct_a
+      task ? task.array_index : 0
```

Comment:
It looks like slurm uses `2**32 - 2` for non-array jobs here.  I'm not sure how consistently it uses that number across platforms and architectures, but it is consistent on the single machine I've tested it.

I can't think of any good reason for us to copy that.  `0` seems like a suitably valid value, and I think it unlikely that anyone is relying on `%a` expanding to `2**32 - 2` for non-array jobs. Probably.

Reply:
Hmm `2**32 -2` looks like an odd number. My guess it's "-1" as an unsigned 32-bit int

Reply:
Yep! It's an algorithmic underflow :laughing:
```
2**32-2 = 0xFFFFFFFE
```